src/proof.rs

```diff
@@ -197,6 +204,71 @@ impl Lemma {
         }
     }
 
+    /// Attempts to generate a proof that the `idx`-th leaf is a member of
+    /// the given tree. The `count` must equal the number of leaves in the
+    /// `tree`. If `idx >= count`, `None` is returned. Otherwise it returns
+    /// the new `Lemma` and the `idx`-th value.
+    pub fn new_by_index<T>(tree: &Tree<T>, idx: usize, count: usize) -> Option<(Lemma, &T)> {
+        if idx >= count {
+            return None;
+        }
+        match *tree {
+            Tree::Empty { .. } => None,
+
+            Tree::Leaf {
+                ref hash,
+                ref value,
+                ..
+            } => {
+                if count != 1 {
+                    return None;
+                }
+                let lemma = Lemma {
+                    node_hash: hash.clone(),
+                    sibling_hash: None,
+                    sub_lemma: None,
+                };
+                Some((lemma, value))
+            }
+
+            Tree::Node {
+                ref hash,
+                ref left,
+                ref right,
+            } => {
+                let left_count = count.next_power_of_two() / 2;
+                let (sub_lem_val, sibling_hash);
+                if idx < left_count {
+                    sub_lem_val = Lemma::new_by_index(left, idx, left_count);
+                    sibling_hash = Positioned::Right(right.hash().clone());
+                } else {
+                    sub_lem_val = Lemma::new_by_index(right, idx - left_count, count - left_count);
+                    sibling_hash = Positioned::Left(left.hash().clone());
+                }
+                sub_lem_val.map(|(sub_lemma, value)| {
+                    let lemma = Lemma {
+                        node_hash: hash.clone(),
+                        sibling_hash: Some(sibling_hash),
+                        sub_lemma: Some(Box::new(sub_lemma)),
+                    };
+                    (lemma, value)
+                })
+            }
+        }
+    }
+
+    /// Returns the index of this lemma's value, given the total number of items in the tree.
+    pub fn index(&self, count: usize) -> usize {
+        let left_count = count.next_power_of_two() / 2;
+        match (self.sub_lemma.as_ref(), self.sibling_hash.as_ref()) {
```

Comment:
When is that `(None, Some(_))` will ever match here? In `new_by_index` every `Lemma` created will match `(Some(_), Some(_))` ([here](https://github.com/SpinResearch/merkle.rs/pull/36/files#diff-167565242b5fe3236877e92612bd630bR249)) or `(None, None)` ([here](https://github.com/SpinResearch/merkle.rs/pull/36/files#diff-167565242b5fe3236877e92612bd630bR249)). Am I missing something again?

Reply:
I considered whether it should return a Result for a little bit, but I think it's a fair assumption that users won't manually edit `Lemma`s, and thus something is wrong with our code if it ever hits these conditions, so we should panic.

Reply:
Actually, now that I think of it, I think panicking here would be dangerous: If this is used in any networking application, an attacker could send us a `Lemma` that breaks this invariant over the wire. That would successfully deserialize and then cause our node to panic here.

If these combinations really don't make sense, that should be reflected in the type itself, e.g.:

```rust
pub struct Lemma {
    pub node_hash: Vec<u8>,
    pub sub_lemma: Option<(Positioned<Vec<u8>>, Box<Lemma>)>,
}
```

If you prefer to keep the current `Lemma` structure, we should instead make sure that the invariant (either both `None` or both `Some`) cannot be broken: We should adapt the serde and protobuf deserialization methods so that they return an error on `None`+`Some`, and add the invariant to the type's documentation.

Reply:
Good point. How about we do panic, but in the docstring for the method we make that clear, and tell the user they should run `Proof::validate` on the input before ever running `Proof::index`?

I do like as much as possible tying being an object of a certain type with that object being valid, but there's limits to this when writing in anything that isn't a formal proving language. I think the suggestion to restructure `Lemma` would be appropriate for an issue, some further discussion, and perhaps a followup PR. For now, however, I think we should keep this scoped and worked with the structs there already are.